-- MySQL dump 10.13  Distrib 5.6.33, for debian-linux-gnu (x86_64)
--
-- Host: mysql-vm-022.ebi.ac.uk    Database: emg
-- ------------------------------------------------------
-- Server version	5.6.33-log

--
-- Dumping data for table `VARIABLE_NAMES`
--
LOCK TABLES `VARIABLE_NAMES` WRITE;
/*!40000 ALTER TABLE `VARIABLE_NAMES` DISABLE KEYS */;
INSERT INTO `VARIABLE_NAMES` VALUES (1,'sample storage temperature','temperature at which sample was stored, e.g. -80','{float} {unit}','samp_store_temp','GSC_v4.0','sample','n','n','air;host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;microbial mat/biofilm;miscellaneous natural or artificial environment;plant-associated;sediment;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (2,'sex','physical sex of the host','[male|female|neuter|hermaphrodite|not determined]','sex','GSC_v4.0','sample','n','n','host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (3,'silicate','concentration of silicate','{float} {unit}','silicate','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (4,'sodium','sodium concentration','{float} {unit}','sodium','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (5,'sulfate','concentration of sulfate','{float} {unit}','sulfate','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (6,'sulfide','concentration of sulfide','{float} {unit}','sulfide','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (7,'temperature','temperature of the sample at time of sampling','{float} {unit}','temp','GSC_v4.0','sample','n','n','air;host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;microbial mat/biofilm;miscellaneous natural or artificial environment;plant-associated;sediment;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (8,'tidal stage','stage of tide','[low|high]','tidal_stage','GSC_v4.0','sample','n','n','sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (9,'total carbon','total carbon content','{float} {unit}','tot_carb','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (10,'total mass','total mass of the host at collection, the unit depends on host','{float} {unit}','tot_mass','GSC_v4.0','sample','n','n','host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;plant-associated',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (11,'total nitrogen','total nitrogen concentration, calculated by: total nitrogen = total dissolved nitrogen + particulate nitrogen. Can also be measured without filtering, reported as nitrogen','{float} {unit}','tot_nitro','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (12,'total organic carbon','Definition for soil: total organic C content of the soil units of g C/kg soil. Definition otherwise: total organic carbon content','{float} {unit}','tot_org_carb','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;soil',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (13,'turbidity','turbidity measurement','{float} {unit}','turbidity','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (14,'water content','water content measurement','{float} {unit}','water_content','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (15,'submitted to insdc','Depending on the study (large-scale e.g. done with next generation sequencing technology, or small-scale) sequences have to be submitted to SRA (Sequence Read Archive), DRA (DDBJ Read Archive) or via the classical Webin/Sequin systems to Genbank, ENA and DDBJ','{boolean}','submitted_to_insdc','GSC_v2.1','none','y','y','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (16,'investigation type','Nucleic Acid Sequence Report is the root element of all MIGS/MIMS compliant reports as standardized by Genomic Standards Consortium. This field is either eukaryote,bacteria,virus,plasmid,organelle, metagenome, mimarks-survey or mimarks-culture','[eukaryote|bacteria_archaea|plasmid|virus|organelle|metagenome,mimarks-survey|mimarks-culture]','investigation_type','GSC_v4.0','sample','y','y','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (17,'project name','Name of the project within which the sequencing was organized','{text}','project_name','GSC_v4.0','study','y','y','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (18,'experimental factor','Experimental factors are essentially the variable aspects of an experiment design which can be used to describe an experiment, or set of experiments, in an increasingly detailed manner. This field accepts ontology terms from Experimental Factor Ontology (EFO) and/or Ontology for Biomedical Investigations (OBI). For a browser of EFO (v132) terms, please see http://bioportal.bioontology.org/visualize/44641; for a browser of OBI (v1.0) terms please see http://bioportal.bioontology.org/visualize/40832','{term|text}','experimental_factor','GSC_v4.0','study','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (19,'geographic location (longitude)','The longitudinal part of the geographical origin of the sample. The values should be reported in decimal degrees and in WGS84 system\n','{float} {float}','lat_lon','GSC_v4.0','sample','y','y','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (20,'geographic location (depth)','Please refer to the definitions of depth in the environmental packages','{float} {unit}','depth','GSC_v2.1','sample','s','s','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (21,'geographic location (altitude)','Please refer to the definitions of either altitude in the environmental packages','{float} {unit}','alt_elev','GSC_v4.0','sample','s','s','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (22,'geographic location (country and/or sea,region)','The geographical origin of the sample as defined by the country or sea name followed by specific region name. Country or sea names should be chosen from the INSDC country list (http://insdc.org/country.html), or the GAZ ontology (v1.446) (http://bioportal.bioontology.org/visualize/40651)','{term}:{term}:{text}','geo_loc_name','GSC_v4.0','sample','y','y','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (23,'collection date','The time of sampling, either as an instance (single point in time) or interval. In case no exact time is available, the date/time can be right truncated i.e. all of these are valid times: 2008-01-23T19:23:10+00:00; 2008-01-23T19:23:10; 2008-01-23; 2008-01; 2008; Except: 2008-01; 2008 all are ISO8601 compliant','{timestamp}','collection_date','GSC_v4.0','sample','y','y','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (24,'environment (biome)','In environmental biome level are the major classes of ecologically similar communities of plants, animals, and other organisms. Biomes are defined based on factors such as plant structures, leaf types, plant spacing, and other factors like climate. Examples include: desert, taiga, deciduous woodland, or coral reef. EnvO (v1.53) terms listed under environmental biome can be found from the link: http://bioportal.bioontology.org/visualize/44405/?conceptid=ENVO%3A00000428','{term}','biome','GSC_v4.0','sample','y','y','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (25,'environment (feature)','Environmental feature level includes geographic environmental features. Examples include: harbor, cliff, or lake. EnvO (v1.53) terms listed under environmental feature can be found from the link: http://bioportal.bioontology.org/visualize/44405/?conceptid=ENVO%3A00002297','{term}','feature','GSC_v4.0','sample','y','y','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (26,'environment (material)','The environmental material level refers to the matter that was displaced by the sample, prior to the sampling event. Environmental matter terms are generally mass nouns. Examples include: air, soil, or water. EnvO (v1.53) terms listed under environmental matter can be found from the link: http://bioportal.bioontology.org/visualize/44405/?conceptid=ENVO%3A00010483','{term}','material','GSC_v4.0','sample','y','y','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (27,'environmental package','MIGS/MIMS/MIMARKS extension for reporting of measurements and observations obtained from one or more of the environments where the sample was obtained. All environmental packages listed here are further defined in separate subtables. By giving the name of the environmental package, a selection of fields can be made from the subtables and can be reported','[air|host-associated|human-associated|human-skin|human-oral|human-gut|human-vaginal|microbial mat/biofilm|miscellaneous natural or artificial environment|plant-associated|sediment|soil|wastewater/sludge|water]','env_package','GSC_v4.0','sample','y','y','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (28,'source material identifiers','The name of the culture collection, holder of the voucher or an institution. Could enumerate a list of common resources, just as the American Type Culture Collection (ATCC), German Collection of Microorganisms and Cell Cultures (DSMZ) etc. Can select not deposited','{PMID|DOI|URL|text}','source_mat_id','GSC_v4.0','sample','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (29,'specific host','If there is a host involved, please provide its taxid (or environmental if not actually isolated from the dead or alive host - i.e. pathogen could be isolated from a swipe of a bench etc) and report whether it is a laboratory or natural host). From this we can calculate any number of groupings of hosts (e.g. animal vs plant, all fish hosts, etc)','host taxid, unknown, environmental','specific_host','GSC_v4.0','sample','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (30,'relationship to oxygen','Is this organism an aerobe, anaerobe? Please note that aerobic and anaerobic are valid descriptors for microbial environments','[aerobe|anaerobe|facultative|microaerophilic|microanaerobe|obligate aerobe|obligate anaerobe]','rel_to_oxygen','GSC_v4.0','sample','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (31,'sample collection device or method','The method or deviced employed for collecting the sample','{text}','samp_collect_device','GSC_v4.0','sample','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (32,'sample material processing','Any processing applied to the sample during or after retrieving the sample from environment. This field accepts OBI, for a browser of OBI (v1.0) terms please see http://bioportal.bioontology.org/visualize/40832','{text|term}','samp_mat_process','GSC_v4.0','sample','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (33,'amount or size of sample collected','Amount or size of sample (volume, mass or area) that was collected','{float} {unit}','samp_size','GSC_v4.0','sample','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (34,'nucleic acid extraction','Link to a literature reference, electronic resource or a standard operating procedure (SOP)','{PMID|DOI|URL}','nucl_acid_ext','GSC_v4.0','sample','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (35,'nucleic acid amplification','Link to a literature reference, electronic resource or a standard operating procedure (SOP)','{PMID|DOI|URL}','nucl_acid_amp','GSC_v4.0','sample','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (36,'library size','Total number of clones in the library prepared for the project','{integer}','lib_size','GSC_v4.0','experiment','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (37,'library reads sequenced','Total number of clones sequenced from the library','{integer}','lib_reads_seqd','GSC_v4.0','experiment','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (38,'library construction method','Library construction method used for clone libraries','{text}','lib_const_meth','GSC_v4.0','experiment','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (39,'library vector','Cloning vector type(s) used in construction of libraries','{text}','lib_vector','GSC_v4.0','experiment','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (40,'library screening strategy','Specific enrichment or screening methods applied before and/or after creating clone libraries in order to select a specific group of sequences','{text}','lib_screen','GSC_v4.0','experiment','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (41,'target gene','Targeted gene or locus name for marker gene studies','{text}','target_gene','GSC_v4.0','experiment','y','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (42,'target subfragment','Name of subfragment of a gene or locus. Important to e.g. identify special regions on marker genes like V6 on 16S rRNA','{text}','target_subfragment','GSC_v4.0','experiment','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (43,'pcr primers','PCR primers that were used to amplify the sequence of the targeted gene, locus or subfragment. This field should contain all the primers used for a single PCR reaction if multiple forward or reverse primers are present in a single PCR reaction. The primer sequence should be reported in uppercase letters','FWD:{dna} REV:{dna}','pcr_primers','GSC_v4.0','experiment','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (44,'multiplex identifiers','Molecular barcodes, called Multiplex Identifiers (MIDs), that are used to specifically tag unique samples in a sequencing run. Sequence should be reported in uppercase letters','{dna}','mid','GSC_v4.0','experiment','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (45,'adapters','Adapters provide priming sequences for both amplification and sequencing of the sample-library fragments. Both adapters should be reported; in uppercase letters','{dna},{dna}','adapters','GSC_v4.0','experiment','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (46,'pcr conditions','Description of reaction conditions and components for PCR in the form of  \'initial denaturation:94degC_1.5min; annealing=...\'','initial denaturation:degrees_minutes; annealing:degrees_minutes; elongation: degrees_minutes; final elongation:degrees_minutes; total cycles','pcr_cond','GSC_v4.0','experiment','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (47,'sequencing method','Sequencing method used; e.g. Sanger, pyrosequencing, ABI-solid','{text}','seq_meth','GSC_v4.0','experiment','y','y','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (48,'sequence quality check','Indicate if the sequence has been called by automatic systems (none) or undergone a manual editing procedure (e.g. by inspecting the raw data or chromatograms). Applied only for sequences that are not submitted to SRA or DRA','[none|manually edited]','seq_quality_check','GSC_v4.0','analysis','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (49,'chimera check','A chimeric sequence, or chimera for short, is a sequence comprised of two or more phylogenetically distinct parent sequences. Chimeras are usually PCR artifacts thought to occur when a prematurely terminated amplicon reanneals to a foreign DNA strand and is copied to completion in the following PCR cycles. The point at which the chimeric sequence changes from one parent to the next is called the breakpoint or conversion point','{text} {text}','chimera_check','GSC_v4.0','analysis','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (50,'assembly','How was the assembly done (e.g. with a text based assembler like phrap or a flowgram assembler); estimated error rate associated with the finished sequences (e.g. error rate of 1 in 1000 bp); and the method of calculation. E.g.{assembly method}; {estimated error rate}; {method of calculation}','{text};{text};{text}','assembly','GSC_v2.1','analysis','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (51,'relevant electronic resources','URL of electronic resources relevant to the sample','{URL}','url','GSC_v4.0','sample','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (52,'relevant standard operating procedures','Standard operating procedures used in assembly and/or annotation of genomes, metagenomes or environmental sequences','{PMID|DOI|URL}','sop','GSC_v4.0','analysis','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (53,'age','age of host at the time of sampling; relevant scale depends on species and study, e.g. could be seconds for amoebae or centuries for trees','{float} {unit}','age','GSC_v4.0','sample','n','n','host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;plant-associated',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (54,'alkalinity','alkalinity, the ability of a solution to neutralize acids to the equivalence point of carbonate or bicarbonate','{float} {unit}','alkalinity','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (55,'alkyl diethers','concentration of alkyl diethers ','{float} {unit}','alkyl_diethers','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (56,'altitude','The altitude of the sample is the vertical distance between Earth\'s surface above Sea Level and the sampled position in the air.','{float} m','alt','GSC_v2.1','sample','s','s','air;host-associated;microbial mat/biofilm;miscellaneous natural or artificial environment;plant-associated;sediment;soil;water','mandatory for air samples');
INSERT INTO `VARIABLE_NAMES` VALUES (57,'aminopeptidase activity','measurement of aminopeptidase activity','{float} {unit}','aminopept_act','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (58,'ammonium','concentration of ammonium','{float} {unit}','ammonium','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (59,'bacterial carbon production','measurement of bacterial carbon production','{float} {unit}','bacteria_carb_prod','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (60,'biomass','amount of biomass; should include the name for the part of biomass measured, e.g. microbial, total. can include multiple measurements','{text};{float} {unit}','biomass','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (61,'bishomohopanol','concentration of bishomohopanol ','{float} {unit}','bishomohopanol','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (62,'body product','substance produced by the body, e.g. stool, mucus, where the sample was obtained from','{text}','body_product','GSC_v4.0','sample','n','n','host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;plant-associated',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (63,'body site','name of body site where the sample was obtained fromhttp://bioportal.bioontology.org/visualize/44507','{text}','body_site','GSC_v4.0','sample','n','n','host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (64,'body-mass index','body mass index, calculated as weight/(height)squared','{float} {unit}','body_mass_index','GSC_v2.1','sample','n','n','human-associated;human-gut;human-oral;human-skin;human-vaginal',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (65,'bromide','concentration of bromide ','{float} {unit}','bromide','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (66,'calcium','concentration of calcium','{float} {unit}','calcium','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (67,'carbon/nitrogen ratio','ratio of amount or concentrations of carbon to nitrogen','{float} {unit}','carb_nitro_ratio','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (68,'chemical administration','list of chemical compounds administered to the host or site where sampling occurred, and when (e.g. antibiotics, N fertilizer, air filter); can include multiple compounds. For Chemical Entities of Biological Interest ontology (CHEBI) (v1.72), please see http://bioportal.bioontology.org/visualize/44603','{term}; {timestamp}','chem_administration','GSC_v4.0','sample','n','n','air;host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;microbial mat/biofilm;miscellaneous natural or artificial environment;plant-associated;sediment;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (69,'chloride','concentration of chloride ','{float} {unit}','chloride','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (70,'chlorophyll','concentration of chlorophyll','{float} {unit}','chlorophyll','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (71,'current','measurement of current ','{float} {unit}','current','GSC_v2.1','sample','n','n','miscellaneous natural or artificial environment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (72,'density','density of sample','{float} {unit}','density','GSC_v4.0','sample','n','n','miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (73,'depth','Depth is defined as the vertical distance below surface, e.g. for sediment or soil samples depth is measured from sediment or soil surface, respectivly. Depth can be reported as an interval for subsurface samples.','{float} m','depth','GSC_v4.0','sample','n','n','host-associated;microbial mat/biofilm;miscellaneous natural or artificial environment;plant-associated;sediment;soil;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (74,'diet','type of diet depending on the host, for animals omnivore, herbivore etc., for humans high-fat, meditteranean etc.; can include multiple diet types','{text}','diet','GSC_v4.0','sample','n','n','host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (75,'diether lipids','concentration of diether lipids; can include multiple types of diether lipids','{text};{float} {unit}','diether_lipids','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (76,'disease status','list of diseases with which the host has been diagnosed; can include multiple diagnoses. the value of the field depends on host; for humans the terms should be chosen from DO (Disease Ontology), other hosts are free text. For DO terms, please see http://gemina.svn.sourceforge.net/viewvc/gemina/trunk/Gemina/ontologies/gemina_symptom.obo?view=log','{term}','disease_stat','GSC_v4.0','sample','n','n','host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;plant-associated',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (77,'dissolved carbon dioxide','concentration of dissolved carbon dioxide','{float} {unit}','diss_carb_dioxide','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (78,'dissolved hydrogen','concentration of dissolved hydrogen','{float} {unit}','diss_hydrogen','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (79,'dissolved inorganic carbon','dissolved inorganic carbon concentration','{float} {unit}','diss_inorg_carb','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (80,'dissolved organic carbon','concentration of dissolved organic carbon','{float} {unit}','diss_org_carb','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (81,'dissolved organic nitrogen','dissolved organic nitrogen concentration measured as; total dissolved nitrogen - NH4 - NO3 - NO2','{float} {unit}','diss_org_nitro','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (82,'dissolved oxygen','concentration of dissolved oxygen','{float} {unit}','diss_oxygen','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (83,'dry mass','measurement of dry mass','{float} {unit}','dry_mass','GSC_v4.0','sample','n','n','host-associated;plant-associated',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (84,'elevation','The elevation of the sampling site as measured by the vertical distance from mean sea level.','{float} {unit}','elev','GSC_v2.1','sample','n','n','air;host-associated;microbial mat/biofilm;miscellaneous natural or artificial environment;plant-associated;sediment;soil;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (85,'family relationship','relationships to other hosts in the same study; can include multiple relationships','{text};{text}','family_relationship','GSC_v4.0','sample','n','n','host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (86,'genotype','observed genotype','{text}','genotype','GSC_v4.0','sample','n','n','host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;plant-associated',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (87,'glucosidase activity','measurement of glucosidase activity','{float} {unit}','glucosidase_act','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (88,'height','the height of subject','{float} {unit}','height','GSC_v2.1','sample','n','n','human-associated;human-gut;human-oral;human-skin;human-vaginal',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (89,'height or length','measurement of height or length','{float} {unit}','height_or_length','GSC_v4.0','sample','n','n','host-associated;plant-associated',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (90,'host body temperature','core body temperature of the host when sample was collected','{float} {unit}','host_body_temp','GSC_v4.0','sample','n','n','host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (91,'host common name','common name of the host, e.g. human','{text}','host_common_name','GSC_v4.0','sample','n','n','host-associated;plant-associated',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (92,'host subject id','a unique identifier by which each subject can be referred to, de-identified, e.g. #131','{text}','host_subject_id','GSC_v4.0','sample','n','n','host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (93,'host taxid','NCBI taxon id of the host, e.g. 9606','{integer}','host_taxid','GSC_v4.0','sample','n','n','host-associated;plant-associated',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (94,'IHMC ethnicity','ethnicity of the subject','{integer|text}','ihmc_ethnicity','GSC_v4.0','sample','n','n','human-associated;human-gut;human-oral;human-skin;human-vaginal',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (95,'IHMC medication code','can include multiple medication codes','{integer}','ihmc_medication_code','GSC_v4.0','sample','n','n','human-associated;human-gut;human-oral;human-skin;human-vaginal',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (96,'infra_specific_name','taxonomic information about the host below subspecies level','{text}','infra_specific_name','GSC_v4.0','sample','n','n','host-associated;plant-associated',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (97,'infra_specific_rank','taxonomic rank information about the host below subspecies level, such as variety, form, rank etc.','{text}','infra_specific_rank','GSC_v4.0','sample','n','n','host-associated;plant-associated',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (98,'last meal','content of last meal and time since feeding; can include multiple values','{text};{period}','last_meal','GSC_v4.0','sample','n','n','host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (99,'life stage','description of life stage of host','{text}','life_stage','GSC_v4.0','sample','n','n','host-associated;plant-associated',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (100,'magnesium','concentration of magnesium','{float} {unit}','magnesium','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (101,'mean friction velocity','measurement of mean friction velocity','{float} {unit}','mean_frict_vel','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (102,'mean peak friction velocity','measurement of mean peak friction velocity','{float} {unit}','mean_peak_frict_vel','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (103,'medical history performed','whether full medical history was collected','{boolean}','medic_hist_perform','GSC_v4.0','sample','n','n','human-associated;human-gut;human-oral;human-skin;human-vaginal',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (104,'methane','methane (gas) amount or concentration at the time of sampling','{float} {unit}','methane','GSC_v4.0','sample','n','n','air;microbial mat/biofilm;sediment',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (105,'miscellaneous parameter','any other measurement performed or parameter collected, that is not listed here','{text};{float} {unit}','misc_param','GSC_v4.0','sample','n','n','air;host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;microbial mat/biofilm;miscellaneous natural or artificial environment;plant-associated;sediment;soil;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (106,'n-alkanes','concentration of n-alkanes; can include multiple n-alkanes','{text};{float} {unit}','n_alkanes','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (107,'nitrate','concentration of nitrate','{float} {unit}','nitrate','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (108,'nitrite','concentration of nitrite','{float} {unit}','nitrite','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (109,'nitrogen','concentration of nitrogen (total)','{float} {unit}','nitro','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (110,'occupation','most frequent job performed by subject','{integer}','occupation','GSC_v2.1','sample','n','n','human-associated;human-gut;human-oral;human-skin;human-vaginal',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (111,'organic carbon','concentration of organic carbon','{float} {unit}','org_carb','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (112,'organic matter','concentration of organic matter ','{float} {unit}','org_matter','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (113,'organic nitrogen','concentration of organic nitrogen','{float} {unit}','org_nitro','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (114,'organism count','total count of any organism per gram or volume of sample,should include name of organism followed by count; can include multiple organism counts','{text};{float} {unit}','organism_count','GSC_v4.0','sample','n','n','air;host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;microbial mat/biofilm;miscellaneous natural or artificial environment;plant-associated;sediment;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (115,'oxygenation status of sample','oxygenation status of sample','[aerobic|anaerobic]','oxy_stat_samp','GSC_v4.0','sample','n','n','air;host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;microbial mat/biofilm;miscellaneous natural or artificial environment;plant-associated;sediment;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (116,'particulate organic carbon','concentration of particulate organic carbon','{float} {unit}','part_org_carb','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (117,'perturbation','type of perturbation, e.g. chemical administration, physical disturbance, etc., coupled with time that perturbation occurred; can include multiple perturbation types','{text};{interval}','perturbation','GSC_v4.0','sample','n','n','air;host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;microbial mat/biofilm;miscellaneous natural or artificial environment;plant-associated;sediment;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (118,'petroleum hydrocarbon','concentration of petroleum hydrocarbon','{float} {unit}','petroleum_hydrocarb','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (119,'pH','pH measurement','{float} {unit}','ph','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;soil;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (120,'phaeopigments','concentration of phaeopigments; can include multiple phaeopigments','{text};{float} {unit}','phaeopigments','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (121,'phenotype','phenotype of host. For Phenotypic quality Ontology (PATO) (v1.269) terms, please see http://bioportal.bioontology.org/visualize/44601','{term}','phenotype','GSC_v4.0','sample','n','n','host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;plant-associated',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (122,'phosphate','concentration of phosphate','{float} {unit}','phosphate','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (123,'phospholipid fatty acid','concentration of phospholipid fatty acids; can include multiple values','{text};{float} {unit}','phosplipid_fatt_acid','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (124,'potassium','concentration of potassium ','{float} {unit}','potassium','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (125,'pressure','pressure to which the sample is subject, in atmospheres','{float} {unit}','pressure','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (126,'pulse','resting pulse, measured as beats per minute','{float} {unit}','pulse','GSC_v2.1','sample','n','n','human-associated;human-gut;human-oral;human-skin;human-vaginal',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (127,'redox potential','redox potential, measured relative to a hydrogen cell, indicating oxidation or reduction potential','{float} {unit}','redox_potential','GSC_v4.0','sample','n','n','microbial mat/biofilm;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (128,'salinity','salinity measurement','{float} {unit}','salinity','GSC_v4.0','sample','n','n','microbial mat/biofilm;miscellaneous natural or artificial environment;sediment;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (129,'sample salinity','salinity of sample, i.e. measure of total salt concentration','{float} {unit}','samp_salinity','GSC_v4.0','sample','n','n','air;host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;plant-associated;wastewater/sludge',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (130,'sample storage duration','duration for which sample was stored','{interval}','samp_store_dur','GSC_v4.0','sample','n','n','air;host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;microbial mat/biofilm;miscellaneous natural or artificial environment;plant-associated;sediment;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (131,'sample storage location','location at which sample was stored, usually name of a specific freezer/room','{text}','samp_store_loc','GSC_v4.0','sample','n','n','air;host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;microbial mat/biofilm;miscellaneous natural or artificial environment;plant-associated;sediment;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (132,'geographic location (elevation)','Please refer to the definitions of either elevation in the environmental packages','{float} {unit}','alt_elev','GSC_v4.0','sample','s','s','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (133,'geographic location (latitude)','The latitudinal part of the geographical origin of the sample. The values should be reported in decimal degrees and in WGS84 system\n','{float} {float}','lat_lon','GSC_v4.0','sample','y','y','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (134,'total depth of water column','measurement of total depth of water column','{float} {unit}','tot_depth_water_col','GSC_v4.0','sample','n','n','water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (135,'plant body site','part of plant from which sample was taken, eg. Root, flower, leaf etc…','{text}','plant_body_site','GSC_v4.0','sample','n','n','n/a','added for Phil Pooles data');
INSERT INTO `VARIABLE_NAMES` VALUES (136,'alternative identifiers','Names or identifiers used in other resources to represent this sample, multiple values allowed. {Source}:{value};{source}:{value} etc…','{text}:{text};',NULL,'non-gsc','sample','n','n','n/a','added for gap and geo IDs');
INSERT INTO `VARIABLE_NAMES` VALUES (137,'miscellaneous parameter2','Any other measurement performed or parameter collected, that is not listed here','{text};{float} {unit}','misc_param','GSC_v2.1','sample','n','n','air;host-associated;human-associated;human-gut;human-oral;human-skin;human-vaginal;microbial mat/biofilm;miscellaneous natural or artificial environment;plant-associated;sediment;soil;wastewater/sludge;water',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (138,'miscellaneous parameter3',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (139,'diet-breastmilk','Does the infant diet include berastmilk, yes or no?','[yes|no]','BreastMilk','non-gsc','sample','n','n','n/a','added for Developing infant gut microbiome project.');
INSERT INTO `VARIABLE_NAMES` VALUES (140,'diet-formula','Does the infant diet include formula, yes or no?','[yes|no]','Formula','non-gsc','sample','n','n','n/a','added for Developing infant gut microbiome project.');
INSERT INTO `VARIABLE_NAMES` VALUES (141,'diet-solids','Does the infant diet include solids, yes or no? Use Pre for infants not started on solids yet, post if they have.','[yes|no|pre|post]','Solids','non-gsc','sample','n','n','n/a','added for Developing infant gut microbiome project.');
INSERT INTO `VARIABLE_NAMES` VALUES (142,'diet-cowsmilk','Does the infant diet include cowsmilk, yes or no?','[yes|no]','CowsMilk','non-gsc','sample','n','n','n/a','added for Developing infant gut microbiome project.');
INSERT INTO `VARIABLE_NAMES` VALUES (143,'alt id - dbgap sample id','Names or identifiers used in dbGAP to represent this sample.','{text}',NULL,'non-gsc','sample','n','n','n/a','added for gap sample IDs');
INSERT INTO `VARIABLE_NAMES` VALUES (144,'alt id - dbgap subject id','Names or identifiers used in dbGAP to represent the subject from which this sample was obtained.','{text}',NULL,'non-gsc','sample','n','n','n/a','added for gap subject IDs');
INSERT INTO `VARIABLE_NAMES` VALUES (145,'alt id - submitted sample id','Name or identifier used in by submitter to represent this sample','{text}',NULL,'non-gsc','sample','n','n','n/a','added for additional sample ID information as used by the submitter prior to accessioning');
INSERT INTO `VARIABLE_NAMES` VALUES (146,'alt id - submitted subject id','Name or identifier used in by submitter to represent the subject from which this sample was obtained','{text}',NULL,'non-gsc','sample','n','n','n/a','added for additional Subject ID information as used by the submitter prior to accessioning');
INSERT INTO `VARIABLE_NAMES` VALUES (147,'air temperature regimen','Description of the air temperature regimen maintained in a glass-house/controlled environment (usually for plant growing)','{text}','air_temp_regm','GSC_v4.0','sample','n','n','air;microbial mat/biofilm;miscellaneous natural or artificial environment;plant-associated;','added for Phil Pooles data');
INSERT INTO `VARIABLE_NAMES` VALUES (148,'faecal colour description','Description of the colour of the faecal matter collected for sampling','{text}',NULL,'non-gsc','sample','n','n','host-associated;human-associated;human-gut','added for Developing infant gut microbiome project.');
INSERT INTO `VARIABLE_NAMES` VALUES (149,'climate environment','treatment involving an exposure to a particular climate; can include multiple climates','{text}','climate_environment','GSC_v4.0','sample','n','n',NULL,'added for Phil Pooles data - unsure what this data point actually means?!');
INSERT INTO `VARIABLE_NAMES` VALUES (150,'watering regimen','Description of the watering regimen maintained in a glass-house/controlled environment (usually for plant growing)','{text}','watering_regm','GSC_v4.0','sample','n','n','air;microbial mat/biofilm;miscellaneous natural or artificial environment;plant-associated;','added for Phil Pooles data');
INSERT INTO `VARIABLE_NAMES` VALUES (153,'original repository',NULL,'{text}',NULL,'non-gsc','sample','n','n','n/a','added for gap subject IDs');
INSERT INTO `VARIABLE_NAMES` VALUES (154,'alt id - MG-RAST','Identifier used in MG-RAST server to represent this sample.','nnnnnnn.n','MG-RAST_ID','non-gsc','sample','n','n','n/a','added for db xref to MGRAST');
INSERT INTO `VARIABLE_NAMES` VALUES (155,'time zone','Identify the time zone in which the sample was taken.',NULL,'time_zone','non-gsc','sample','n','n','n/a','added for srs000447sample ');
INSERT INTO `VARIABLE_NAMES` VALUES (156,'alt id - GOLD','Identifier used by Genomes OnLine Database (GOLD) to represent this sample.',NULL,'gold_sample_id','non-gsc','sample','n','n','n/a','added for db xref to GOLD');
INSERT INTO `VARIABLE_NAMES` VALUES (157,'Administrative Contact name','The name of the person(s) to contact about this sample','{text}',NULL,'non-gsc','sample','n','n','n/a',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (158,'Administrative Contact email','The email of the named person(s) to contact about this sample','{text}',NULL,'non-gsc','sample','n','n','n/a',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (159,'Administrative Contact organisation','The organisation responsible for the initial collection and use of this sample and/or the organisation of the adminstrative contact person.','{text}',NULL,'non-gsc','sample','n','n','n/a',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (161,'volume of sample','Not all of the sample that’s collected gets used for DNA extraction, this value gives the volume of sample used for DNA extraction','ml',NULL,'non-gsc','sample','n','n','miscellaneous natural or artificial environment','added for Kimchi fermentation dataset');
INSERT INTO `VARIABLE_NAMES` VALUES (162,'fermentation - period','The period of time that the fermentation process had been going prior to sampling event','days',NULL,'non-gsc','sample','n','n','miscellaneous natural or artificial environment','added for Kimchi fermentation dataset');
INSERT INTO `VARIABLE_NAMES` VALUES (163,'fermentation - temperature','constant temperature that fermentation has been proceeding at','degrees celcius',NULL,'non-gsc','sample','n','n','miscellaneous natural or artificial environment','added for Kimchi fermentation dataset');
INSERT INTO `VARIABLE_NAMES` VALUES (164,'food preparation - method','Method of preparation of food, including things like slicing, mashing, boiling, blanching, freezing etc…','{text}',NULL,'non-gsc','sample','n','n','miscellaneous natural or artificial environment','added for Kimchi fermentation dataset');
INSERT INTO `VARIABLE_NAMES` VALUES (165,'food preparation - ingredients','Ingrediants (including ammounts if possible) used in the prepartion of the food stuff being sampled','{text}',NULL,'non-gsc','sample','n','n','miscellaneous natural or artificial environment','added for Kimchi fermentation dataset');
INSERT INTO `VARIABLE_NAMES` VALUES (166,'Concentration of glucose','Concentration of glucose ','mM',NULL,'non-gsc','sample','n','n','any','added for Kimchi fermentation dataset');
INSERT INTO `VARIABLE_NAMES` VALUES (167,'Concentration of fructose','Concentration of fructose ','mM',NULL,'non-gsc','sample','n','n','any','added for Kimchi fermentation dataset');
INSERT INTO `VARIABLE_NAMES` VALUES (168,'Concentration of sucrose','Concentration of sucrose ','mM',NULL,'non-gsc','sample','n','n','any','added for Kimchi fermentation dataset');
INSERT INTO `VARIABLE_NAMES` VALUES (169,'Concentration of mannitol','Concentration of mannitol ','mM',NULL,'non-gsc','sample','n','n','any','added for Kimchi fermentation dataset');
INSERT INTO `VARIABLE_NAMES` VALUES (170,'Concentration of lactate','Concentration of lactate ','mM',NULL,'non-gsc','sample','n','n','any','added for Kimchi fermentation dataset');
INSERT INTO `VARIABLE_NAMES` VALUES (171,'Concentration of acetate','Concentration of acetate ','mM',NULL,'non-gsc','sample','n','n','any','added for Kimchi fermentation dataset');
INSERT INTO `VARIABLE_NAMES` VALUES (172,'Concentration of ethanol','Concentration of ethanol ','mM',NULL,'non-gsc','sample','n','n','any','added for Kimchi fermentation dataset');
INSERT INTO `VARIABLE_NAMES` VALUES (173,'history/agrochemical additions','addition of fertilizers, pesticides, etc. - amount and time of applications','{text};{float} {unit};{timestamp}','agrochem_addition','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (174,'extreme_unusual_properties/Al saturation','aluminum saturation (esp. for tropical soils)','{float} {unit}','al_sat','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (175,'extreme_unusual_properties/Al saturation method','reference or method used in determining Al saturation','{PMID|DOI|URL}','al_sat_meth','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (176,'amniotic fluid/color','specification of the color of the amniotic fluid sample','{text}','amniotic_fluid_color','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (177,'mean annual and seasonal precipitation','mean annual and seasonal precipitation (mm)','{float} {unit}','annual_season_precpt','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (178,'mean annual and seasonal temperature','mean annual and seasonal temperature (oC)','{float} {unit}','annual_season_temp','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (179,'antibiotic regimen','information about treatment involving antibiotic administration; should include the name of antibiotic, amount administered, treatment duration, interval and total experimental duration; can include multiple antibiotic regimens','{text};{float} {unit};{period};{interval};{period}','antibiotic_regm','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (180,'atmospheric data','measurement of atmospheric data; can include multiple data','{text};{float} {unit}','atmospheric_data','GSC_v4.0','sample','n','n','water',',');
INSERT INTO `VARIABLE_NAMES` VALUES (181,'barometric pressure','force per unit area exerted against a surface by the weight of air above that surface','{float} {unit}','barometric_press','GSC_v4.0','sample','n','n','air',',');
INSERT INTO `VARIABLE_NAMES` VALUES (182,'biochemical oxygen demand','a measure of the relative oxygen-depletion effect of a waste contaminant','{float} {unit}','biochem_oxygen_dem','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (183,'birth control','specification of birth control medication used','{text}','birth_control','GSC_v4.0','sample','n','n','human-vaginal',',');
INSERT INTO `VARIABLE_NAMES` VALUES (184,'blood/blood disorder','history of blood disorders; can include multiple disorders','{text}','blood_blood_disord','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (185,'blood pressure diastolic','resting diastolic blood pressure, measured as mm mercury','{float} {unit}','blood_press_diast','GSC_v4.0','sample','n','n','host-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (186,'blood pressure systolic','resting systolic blood pressure, measured as mm mercury','{float} {unit}','blood_press_syst','GSC_v4.0','sample','n','n','host-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (187,'body habitat','original body habitat where the sample was obtained from','{text}','body_habitat','GSC_v4.0','sample','n','n','host-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (188,'carbon dioxide','carbon dioxide (gas) amount or concentration at the time of sampling','{float} {unit}','carb_dioxide','GSC_v4.0','sample','n','n','air',',');
INSERT INTO `VARIABLE_NAMES` VALUES (189,'carbon monoxide','carbon monoxide (gas) amount or concentration at the time of sampling','{float} {unit}','carb_monoxide','GSC_v4.0','sample','n','n','air',',');
INSERT INTO `VARIABLE_NAMES` VALUES (190,'chemical mutagen','treatment involving use of mutagens; should include the name of mutagen, amount administered, treatment duration, interval and total experimental duration; can include multiple mutagen regimens','{text};{float} {unit};{period};{interval};{period}','chem_mutagen','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (191,'chemical oxygen demand','a measure of the relative oxygen-depletion effect of a waste contaminant','{float} {unit}','chem_oxygen_dem','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (192,'history/crop rotation','whether or not crop is rotated, and if yes, rotation schedule','{boolean};Rn/{timestamp}/{period}','crop_rotation','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (193,'current land use','present state of sample site, choose from this list [cities|farmstead|industrial areas|roads/railroads|rock|sand|gravel|mudflats|salt flats|badlands|permanent snow or ice|saline seeps|mines/quarries|oil waste areas|small grains|row crops|vegetable crops|horticultural plants (e.g. tulips)|marshlands (grass,sedges,rushes)|tundra (mosses,lichens)|rangeland|pastureland (grasslands used for livestock grazing)|hayland|meadows (grasses,alfalfa,fescue,bromegrass,timothy)|shrub land (e.g. mesquite,sage-brush,creosote bush,shrub oak,eucalyptus)|successional shrub land (tree saplings,hazels,sumacs,chokecherry,shrub dogwoods,blackberries)|shrub crops (blueberries,nursery ornamentals,filberts)|vine crops (grapes)|conifers (e.g. pine,spruce,fir,cypress)|hardwoods (e.g. oak,hickory,elm,aspen)|intermixed hardwood and conifers|tropical (e.g. mangrove,palms)|rainforest (evergreen forest receiving >406 cm annual rainfall)|swamp (permanent or semi-permanent water body dominated by woody plants)|crop trees (nuts,fruit,christmas trees,nursery trees)]','{text}','cur_land_use','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (194,'current vegetation','vegetation classification from one or more standard classification systems, or agricultural crop','{text}','cur_vegetation','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (195,'current vegetation method','reference or method used in vegetation classification ','{PMID|DOI|URL}','cur_vegetation_meth','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (196,'dermatology disorder','history of dermatology disorders; can include multiple disorders','{text}','dermatology_disord','GSC_v4.0','sample','n','n','human-skin',',');
INSERT INTO `VARIABLE_NAMES` VALUES (197,'major diet change in last six months','specification of major diet changes in the last six months, if yes the change should be specified','{boolean};{text}','diet_last_six_month','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (198,'dissolved inorganic nitrogen','concentration of dissolved inorganic nitrogen ','{float} {unit}','diss_inorg_nitro','GSC_v4.0','sample','n','n','water',',');
INSERT INTO `VARIABLE_NAMES` VALUES (199,'dissolved inorganic phosphorus','concentration of dissolved inorganic phosphorus ','{float} {unit}','diss_inorg_phosp','GSC_v4.0','sample','n','n','water',',');
INSERT INTO `VARIABLE_NAMES` VALUES (200,'dominant hand','dominant hand of the subject','[left|right|ambidextrous]','dominant_hand','GSC_v4.0','sample','n','n','human-skin',',');
INSERT INTO `VARIABLE_NAMES` VALUES (201,'douche','date of most recent douche','{timestamp}','douche','GSC_v4.0','sample','n','n','human-vaginal',',');
INSERT INTO `VARIABLE_NAMES` VALUES (202,'drainage classification','drainage classification from a standard system such as the USDA system','[very poorly|poorly|somewhat poorly|moderately well|well|excessively drained]','drainage_class','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (203,'drug usage','any drug used by subject and the frequency of usage; can include multiple drugs used','{text};{integer}/[year|month|week|day|hour]','drug_usage','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (204,'efficiency percent','percentage of volatile solids removed from the anaerobic digestor','{float} {unit}','efficiency_percent','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (205,'emulsions','amount or concentration of substances such as paints, adhesives, mayonnaise, hair colorants, emulsified oils, etc.; can include multiple emulsion types','{text};{float} {unit}','emulsions','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (206,'history/extreme events','unusual physical events that may have affected microbial populations','{timestamp}','extreme_event','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (207,'extreme_unusual_properties/salinity','measured salinity ','{float} {unit}','extreme_salinity','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (208,'soil_taxonomic/FAO classification','soil classification from the FAO World Reference Database for Soil Resources','{term}','fao_class','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (209,'fertilizer regimen','information about treatment involving the use of fertilizers; should include the name fertilizer, amount administered, treatment duration, interval and total experimental duration; can include multiple fertilizer regimens','{text};{float} {unit};{period};{interval};{period}','fertilizer_regm','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (210,'history/fire','historical and/or physical evidence of fire','{timestamp}','fire','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (211,'history/flooding','historical and/or physical evidence of flooding','{timestamp}','flooding','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (212,'amniotic fluid/foetal health status','specification of foetal health status, should also include abortion','{text}','foetal_health_stat','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (213,'fungicide regimen','information about treatment involving use of fungicides; should include the name of fungicide, amount administered, treatment duration, interval and total experimental duration; can include multiple fungicide regimens','{text};{float} {unit};{period};{interval};{period}','fungicide_regm','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (214,'gaseous environment','use of conditions with differing gaseous environments; should include the name of gaseous compound, amount administered, treatment duration, interval and total experimental duration; can include multiple gaseous environment regimens','{text};{float} {unit};{period};{interval};{period}','gaseous_environment','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (215,'gaseous substances','amount or concentration of substances such as hydrogen sulfide, carbon dioxide, methane, etc.; can include multiple substances','{text};{float} {unit}','gaseous_substances','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (216,'gastrointestinal tract disorder','history of gastrointestinal tract disorders; can include multiple disorders ','{text}','gastrointest_disord','GSC_v4.0','sample','n','n','human-gut',',');
INSERT INTO `VARIABLE_NAMES` VALUES (217,'amniotic fluid/gestation state','specification of the gestation state','{text}','gestation_state','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (218,'gravidity','whether or not subject is gravid, and if yes date due or date post-conception, specifying which is used','{boolean};{timestamp}','gravidity','GSC_v4.0','sample','n','n','host-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (219,'gravity','information about treatment involving use of gravity factor to study various types of responses in presence, absence or modified levels of gravity; can include multiple treatments','{float} {unit};{period};{interval};{period}','gravity','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (220,'growth hormone regimen','information about treatment involving use of growth hormones; should include the name of growth hormone, amount administered, treatment duration, interval and total experimental duration; can include multiple growth hormone regimens','{text};{float} {unit};{period};{interval};{period}','growth_hormone_regm','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (221,'growth media','information about growth media for growing the plants or tissue cultured samples','[soil|liquid]','growth_med','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (222,'gynecological disorder','history of gynecological disorders; can include multiple disorders','{text}','gynecologic_disord','GSC_v4.0','sample','n','n','human-vaginal',',');
INSERT INTO `VARIABLE_NAMES` VALUES (223,'extreme_unusual_properties/heavy metals','heavy metals present and concentrationsany drug used by subject and the frequency of usage; can include multiple heavy metals and concentrations','{text};{float} {unit}','heavy_metals','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (224,'extreme_unusual_properties/heavy metals method','reference or method used in determining heavy metals','{PMID|DOI|URL}','heavy_metals_meth','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (225,'herbicide regimen','information about treatment involving use of herbicides; information about treatment involving use of growth hormones; should include the name of herbicide, amount administered, treatment duration, interval and total experimental duration; can include multiple regimens','{text};{float} {unit};{period};{interval};{period}','herbicide_regm','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (226,'HIV status','HIV status of subject, if yes HAART initiation status should also be indicated as [YES or NO]','{boolean};{boolean}','hiv_stat','GSC_v2.1','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (227,'horizon','specific layer in the land area which measures parallel to the soil surface and possesses physical characteristics which differ from the layers above and beneath','[O horizon|A horizon|E horizon|B horizon|C horizon|R layer|Permafrost]','horizon','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (228,'horizon method','reference or method used in determining the horizon','{PMID|DOI|URL}','horizon_meth','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (229,'host color','the color of host','{text}','host_color','GSC_v4.0','sample','n','n','host-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (230,'host growth conditions','literature reference giving growth conditions of the host','{PMID|DOI|URL}','host_growth_cond','GSC_v4.0','sample','n','n','host-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (231,'host shape','morphological shape of host ','{text}','host_shape','GSC_v4.0','sample','n','n','host-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (232,'HRT','whether subject had hormone replacement theraphy, and if yes start date','{timestamp}','hrt','GSC_v4.0','sample','n','n','human-vaginal',',');
INSERT INTO `VARIABLE_NAMES` VALUES (233,'humidity','amount of water vapour in the air, at the time of sampling','{float} {unit}','humidity','GSC_v4.0','sample','n','n','air',',');
INSERT INTO `VARIABLE_NAMES` VALUES (234,'humidity regimen','information about treatment involving an exposure to varying degree of humidity; information about treatment involving use of growth hormones; should include amount of humidity administered, treatment duration, interval and total experimental duration; can include multiple regimens','{float} {unit};{period};{interval};{period}','humidity_regm','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (235,'hysterectomy','specification of whether hysterectomy was performed','{boolean}','hysterectomy','GSC_v4.0','sample','n','n','human-vaginal',',');
INSERT INTO `VARIABLE_NAMES` VALUES (236,'industrial effluent percent','percentage of industrial effluents received by wastewater treatment plant','{float} {unit}','indust_eff_percent','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (237,'inorganic particles','concentration of particles such as sand, grit, metal particles, ceramics, etc.; can include multiple particles','{text};{float} {unit}','inorg_particles','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (238,'urine/kidney disorder','history of kidney disorders; can include multiple disorders','{text}','kidney_disord','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (239,'light intensity','measurement of light intensity','{float} {unit}','light_intensity','GSC_v4.0','sample','n','n','water',',');
INSERT INTO `VARIABLE_NAMES` VALUES (240,'links to additional analysis','0','{PMID|DOI|URL}','link_addit_analys','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (241,'link to classification information','link to digitized soil maps or other soil classification information','{PMID|DOI|URL}','link_class_info','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (242,'link to climate information','link to climate resource','{PMID|DOI|URL}','link_climate_info','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (243,'liver disorder','history of liver disorders; can include multiple disorders','{text}','liver_disord','GSC_v4.0','sample','n','n','human-gut',',');
INSERT INTO `VARIABLE_NAMES` VALUES (244,'soil_taxonomic/local classification','soil classification based on local soil classification system','{text}','local_class','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (245,'soil_taxonomic/local classification method','reference or method used in determining the local soil classification ','{PMID|DOI|URL}','local_class_meth','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (246,'amniotic fluid/maternal health status','specification of the maternal health status','{text}','maternal_health_stat','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (247,'mechanical damage','information about any mechanical damage exerted on the plant; can include multiple damages and sites','{text};{text}','mechanical_damage','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (248,'menarche','date of most recent menstruation','{timestamp}','menarche','GSC_v4.0','sample','n','n','human-vaginal',',');
INSERT INTO `VARIABLE_NAMES` VALUES (249,'menopause','date of onset of menopause','{timestamp}','menopause','GSC_v4.0','sample','n','n','human-vaginal',',');
INSERT INTO `VARIABLE_NAMES` VALUES (250,'microbial biomass','the part of the organic matter in the soil that constitutes living microorganisms smaller than 5-10 ��m. IF you keep this, you would need to have correction factors used for conversion to the final units, which should be mg C (or N)/kg soil).','{float} {unit}','microbial_biomass','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (251,'microbial biomass method','reference or method used in determining microbial biomass','{PMID|DOI|URL}','microbial_biomass_meth','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (252,'mineral nutrient regimen','information about treatment involving the use of mineral supplements; should include the name of mineral nutrient, amount administered, treatment duration, interval and total experimental duration; can include multiple mineral nutrient regimens','{text};{float} {unit};{period};{interval};{period}','mineral_nutr_regm','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (253,'non-mineral nutrient regimen','information about treatment involving the exposure of plant to non-mineral nutrient such as oxygen, hydrogen or carbon; should include the name of non-mineral nutrient, amount administered, treatment duration, interval and total experimental duration; can include multiple non-mineral nutrient regimens','{text};{float} {unit};{period};{interval};{period}','non_mineral_nutr_regm','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (254,'nose/mouth/teeth/throat disorder','history of nose/mouth/teeth/throat disorders; can include multiple disorders','{text}','nose_mouth_teeth_throat_disord','GSC_v4.0','sample','n','n','human-oral',',');
INSERT INTO `VARIABLE_NAMES` VALUES (255,'lung/nose-throat disorder','history of nose-throat disorders; can include multiple disorders','{text}','nose_throat_disord','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (256,'organic particles','concentration of particles such as faeces, hairs, food, vomit, paper fibers, plant material, humus, etc.','{text};{float} {unit}','org_particles','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (257,'other','additional relevant information','{text}','other','GSC_v2.1','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (258,'oxygen','oxygen (gas) amount or concentration at the time of sampling','{float} {unit}','oxygen','GSC_v4.0','sample','n','n','air',',');
INSERT INTO `VARIABLE_NAMES` VALUES (259,'particulate organic nitrogen','concentration of particulate organic nitrogen','{float} {unit}','part_org_nitro','GSC_v4.0','sample','n','n','water',',');
INSERT INTO `VARIABLE_NAMES` VALUES (260,'particle classification','particles are classified, based on their size, into six general categories:clay, silt, sand, gravel, cobbles, and boulders; should include amount of particle preceded by the name of the particle type; can include multiple values','{text};{float} {unit}','particle_class','GSC_v2.1','sample','n','n','sediment',',');
INSERT INTO `VARIABLE_NAMES` VALUES (261,'pesticide regimen','information about treatment involving use of insecticides; should include the name of pesticide, amount administered, treatment duration, interval and total experimental duration; can include multiple pesticide regimens','{text};{float} {unit};{period};{interval};{period}','pesticide_regm','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (262,'presence of pets or farm animals','specification of presence of pets or farm animals in the environment of subject, if yes the animals should be specified; can include multiple animals present','{boolean};{text}','pet_farm_animal','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (263,'pH method','reference or method used in determining pH','{PMID|DOI|URL}','ph_meth','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (264,'pH regimen','information about treatment involving exposure of plants to varying levels of pH of the growth media; can include multiple regimen','{float} {unit};{period};{interval};{period}','ph_regm','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (265,'photon flux','measurement of photon flux','{float} {unit}','photon_flux','GSC_v4.0','sample','n','n','water',',');
INSERT INTO `VARIABLE_NAMES` VALUES (266,'plant product','substance produced by the plant, where the sample was obtained from','{text}','plant_product','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (267,'pollutants','pollutant types and, amount or concentrations measured at the time of sampling; can report multiple pollutants by entering numeric values preceded by name of pollutant','{text};{float} {unit}','pollutants','GSC_v4.0','sample','n','n','air',',');
INSERT INTO `VARIABLE_NAMES` VALUES (268,'pooling of DNA extracts (if done)','were multiple DNA extractions mixed? how many?','{boolean};{float} {unit}','pool_dna_extracts','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (269,'porosity','porosity of deposited sediment is volume of voids divided by the total volume of sample','{float} {unit}','porosity','GSC_v2.1','sample','n','n','sediment',',');
INSERT INTO `VARIABLE_NAMES` VALUES (270,'pre-treatment','the process of pre-treatment removes materials that can be easily collected from the raw wastewater','{text}','pre_treatment','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (271,'pregnancy','date due of pregnancy','{timestamp}','pregnancy','GSC_v4.0','sample','n','n','human-vaginal',',');
INSERT INTO `VARIABLE_NAMES` VALUES (272,'history/previous land use','previous land use and dates','{text};{timestamp}','previous_land_use','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (273,'history/previous land use method','reference or method used in determining previous land use and dates','{PMID|DOI|URL}','previous_land_use_meth','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (274,'primary production','measurement of primary production ','{float} {unit}','primary_prod','GSC_v4.0','sample','n','n','water',',');
INSERT INTO `VARIABLE_NAMES` VALUES (275,'primary treatment','the process to produce both a generally homogeneous liquid capable of being treated biologically and a sludge that can be separately treated or processed','{text}','primary_treatment','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (276,'profile position','cross-sectional position in the hillslope where sample was collected.sample area position in relation to surrounding areas','[summit|shoulder|backslope|footslope|toeslope]','profile_position','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (277,'lung/pulmonary disorder','history of pulmonary disorders; can include multiple disorders','{text}','pulmonary_disord','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (278,'radiation regimen','information about treatment involving exposure of plant or a plant part to a particular radiation regimen; should include the radiation type, amount or intensity administered, treatment duration, interval and total experimental duration; can include multiple radiation regimens','{text};{float} {unit};{period};{interval};{period}','radiation_regm','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (279,'rainfall regimen','information about treatment involving an exposure to a given amount of rainfall; can include multiple regimens','{float} {unit};{period};{interval};{period}','rainfall_regm','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (280,'reactor type','anaerobic digesters can be designed and engineered to operate using a number of different process configurations, as batch or continuous, mesophilic, high solid or low solid, and single stage or multistage','{text}','reactor_type','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (281,'respirable particulate matter','concentration of substances that remain suspended in the air, and comprise mixtures of organic and inorganic substances (PM10 and PM2.5); can report multiple PM\'s by entering numeric values preceded by name of PM','{text};{float} {unit}','resp_part_matter','GSC_v4.0','sample','n','n','air',',');
INSERT INTO `VARIABLE_NAMES` VALUES (282,'extreme_unusual_properties/salinity method','reference or method used in determining salinity','{PMID|DOI|URL}','salinity_meth','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (283,'salt regimen','information about treatment involving use of salts as supplement to liquid and soil growth media; should include the name of salt, amount administered, treatment duration, interval and total experimental duration; can include multiple salt regimens','{text};{float} {unit};{period};{interval};{period}','salt_regm','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (284,'sample weight for DNA extraction','weight (g) of soil processed','{float} g','samp_weight_dna_ext','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (285,'seasonal environment','treatment involving an exposure to a particular season (e.g. winter, summer, rabi, rainy etc.)','{text};{period};{interval};{period}','season_environment','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (286,'secondary treatment','the process for substantially degrading the biological content of the sewage ','{text}','secondary_treatment','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (287,'sediment type','information about the sediment type based on major constituents','[biogenous|cosmogenous|hydrogenous|lithogenous]','sediment_type','GSC_v2.1','sample','n','n','sediment',',');
INSERT INTO `VARIABLE_NAMES` VALUES (288,'sewage type','type of wastewater treatment plant as municipial or industrial','{text}','sewage_type','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (289,'sexual activity','current sexual partner and frequency of sex','{text}','sexual_act','GSC_v4.0','sample','n','n','human-vaginal',',');
INSERT INTO `VARIABLE_NAMES` VALUES (290,'composite design/sieving (if any)','collection design of pooled samples and/or sieve size and amount of sample sieved ','{{text}|{float} {unit}};{float} {unit}','sieving','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (291,'slope aspect','the direction a slope faces. While looking down a slope use a compass to record the direction you are facing (direction or degrees); e.g., NW or 315��.  This measure provides an indication of sun and wind exposure that will influence soil temperature and evapotranspiration.','{float} {unit}','slope_aspect','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (292,'slope gradient','commonly called slope.  The angle between ground surface and a horizontal line (in percent).  This is the direction that overland water would flow.  This measure is usually taken with a hand level meter or clinometer. ','{float} {unit}','slope_gradient','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (293,'sludge retention time','the time activated sludge remains in reactor','{float} {unit}','sludge_retent_time','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (294,'smoker','specification of smoking status','{boolean}','smoker','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (295,'soil type','soil series name or other lower-level classification','{text}','soil_type','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (296,'soil type method','reference or method used in determining soil series name or other lower-level classification','{PMID|DOI|URL}','soil_type_meth','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (297,'solar irradiance','the amount of solar energy that arrives at a specific area of a surface during a specific time interval','{float} {unit}','solar_irradiance','GSC_v4.0','sample','n','n','air',',');
INSERT INTO `VARIABLE_NAMES` VALUES (298,'soluble inorganic material','concentration of substances such as ammonia, road-salt, sea-salt, cyanide, hydrogen sulfide, thiocyanates, thiosulfates, etc.','{text};{float} {unit}','soluble_inorg_mat','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (299,'soluble organic material','concentration of substances such as urea, fruit sugars, soluble proteins, drugs, pharmaceuticals, etc.','{text};{float} {unit}','soluble_org_mat','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (300,'soluble reactive phosphorus','concentration of soluble reactive phosphorus','{float} {unit}','soluble_react_phosp','GSC_v4.0','sample','n','n','water',',');
INSERT INTO `VARIABLE_NAMES` VALUES (301,'special diet','specification of special diet; can include multiple special diets','[low carb|reduced calorie|vegetarian|other(to be specified)]','special_diet','GSC_v4.0','sample','n','n','human-gut',',');
INSERT INTO `VARIABLE_NAMES` VALUES (302,'standing water regimen','treatment involving an exposure to standing water during a plant\'s life span, types can be flood water or standing water; can include multiple regimens','{text};{period};{interval};{period}','standing_water_regm','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (303,'storage conditions (fresh/frozen/other)','explain how and for how long the soil sample was stored before DNA extraction.','{text};{period}','store_cond','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (304,'study completion status','specification of study completion status, if no the reason should be specified','[YES|NO-adverse event|NO-non-compliance|NO-lost to follow up|NO-other-specify]','study_complt_stat','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (305,'substrate','the growth substrate of the host ','{text}','substrate','GSC_v4.0','sample','n','n','host-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (306,'suspended particulate matter','concentration of suspended particulate matter','{float} {unit}','suspend_part_matter','GSC_v4.0','sample','n','n','water',',');
INSERT INTO `VARIABLE_NAMES` VALUES (307,'suspended solids','concentration of substances including a wide variety of material, such as silt, decaying plant and animal matter, etc,; can include multiple substances','{text};{float} {unit}','suspend_solids','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (308,'tertiary treatment','the process providing a final treatment stage to raise the effluent quality before it is discharged to the receiving environment','{text}','tertiary_treatment','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (309,'texture','the relative proportion of different grain sizes of mineral particles in a soil, as described using a standard system; express as % sand (50 um to 2 mm), silt (2 um to 50 um), and clay (<2 um) with textural name (e.g., silty clay loam) optional.','{float} {unit}','texture','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (310,'texture method','reference or method used in determining soil texture','{PMID|DOI|URL}','texture_meth','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (311,'history/tillage','note method(s) used for tilling','[drill|cutting disc|ridge till|strip tillage|zonal tillage|chisel|tined|mouldboard|disc plough]','tillage','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (312,'time since last toothbrushing','specification of the time since last toothbrushing','{timestamp}','time_last_toothbrush','GSC_v4.0','sample','n','n','human-oral',',');
INSERT INTO `VARIABLE_NAMES` VALUES (313,'time since last wash','specification of the time since last wash','{timestamp}','time_since_last_wash','GSC_v4.0','sample','n','n','human-skin',',');
INSERT INTO `VARIABLE_NAMES` VALUES (314,'tissue culture growth media','description of plant tissue culture growth media used','{PMID|DOI|URL}','tiss_cult_growth_med','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (315,'total dissolved nitrogen','total dissolved nitrogen concentration, reported as nitrogen, measured by: total dissolved nitrogen = NH4 + NO3NO2 + dissolved organic nitrogen','{float} {unit}','tot_diss_nitro','GSC_v4.0','sample','n','n','water',',');
INSERT INTO `VARIABLE_NAMES` VALUES (316,'total inorganic nitrogen','total inorganic nitrogen content','{float} {unit}','tot_inorg_nitro','GSC_v4.0','sample','n','n','water',',');
INSERT INTO `VARIABLE_NAMES` VALUES (317,'total N','total nitrogen content of the soil Units of g N/kg soil','{float} {unit}','tot_n','GSC_v2.1','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (318,'total N method','reference or method used in determining the total N','{PMID|DOI|URL}','tot_n_meth','GSC_v2.1','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (319,'total organic C method','reference or method used in determining total organic C','{PMID|DOI|URL}','tot_org_c_meth','GSC_v4.0','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (320,'total particulate carbon','total particulate carbon content','{float} {unit}','tot_part_carb','GSC_v4.0','sample','n','n','water',',');
INSERT INTO `VARIABLE_NAMES` VALUES (321,'total phosphorus','total phosphorus concentration, calculated by: total phosphorus = total dissolved phosphorus + particulate phosphorus. Can also be measured without filtering, reported as phosphorus','{float} {unit}','tot_phosp','GSC_v4.0','sample','n','n','water',',');
INSERT INTO `VARIABLE_NAMES` VALUES (322,'total phosphate','total amount or concentration of phosphate','{float} {unit}','tot_phosphate','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (323,'travel outside the country in last six months','specification of the countries travelled in the last six months; can include multiple travels','{text}','travel_out_six_month','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (324,'twin sibling presence','specification of twin sibling presence','{boolean}','twin_sibling','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (325,'urine/collection method','specification of urine collection method','[clean catch|catheter]','urine_collect_meth','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (326,'urogenital disorder','history of urogenital disorders, can include multiple disorders','{text}','urogenit_disord','GSC_v4.0','sample','n','n','human-vaginal',',');
INSERT INTO `VARIABLE_NAMES` VALUES (327,'urine/urogenital tract disorder','history of urogenitaltract disorders; can include multiple disorders','{text}','urogenit_tract_disor','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (328,'ventilation rate','ventilation rate of the system in the sampled premises','{float} {unit}','ventilation_rate','GSC_v4.0','sample','n','n','air',',');
INSERT INTO `VARIABLE_NAMES` VALUES (329,'ventilation type','ventilation system used in the sampled premises','{text}','ventilation_type','GSC_v4.0','sample','n','n','air',',');
INSERT INTO `VARIABLE_NAMES` VALUES (330,'volatile organic compounds','concentration of carbon-based chemicals that easily evaporate at room temperature; can report multiple volatile organic compounds by entering numeric values preceded by name of compound','{text};{float} {unit}','volatile_org_comp','GSC_v4.0','sample','n','n','air',',');
INSERT INTO `VARIABLE_NAMES` VALUES (331,'wastewater type','the origin of wastewater such as human waste, rainfall, storm drains, etc.','{text}','wastewater_type','GSC_v4.0','sample','n','n','wastewater/sludge',',');
INSERT INTO `VARIABLE_NAMES` VALUES (332,'water content of soil','water content (g/g or cm3/cm3)','{float} [g/g|cm3/cm3]','water_content_soil','GSC_v2.1','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (333,'water content of soil method','reference or method used in determining the water content of soil','{PMID|DOI|URL}','water_content_soil_meth','GSC_v2.1','sample','n','n','soil',',');
INSERT INTO `VARIABLE_NAMES` VALUES (334,'water temperature regimen','information about treatment involving an exposure to water with varying degree of temperature; can include multiple regimens','{float} {unit};{period};{interval};{period}','water_temp_regm','GSC_v4.0','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (335,'weight loss in last three months','specification of weight loss in the last three months, if yes should be further specified to include amount of weight loss','{boolean};{float} {unit}','weight_loss_3_month','GSC_v4.0','sample','n','n','human-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (336,'wet mass','measurement of wet mass','{float} {unit}','wet_mass','GSC_v2.1','sample','n','n','plant-associated',',');
INSERT INTO `VARIABLE_NAMES` VALUES (337,'wind direction','wind direction is the direction from which a wind originates','{text}','wind_direction','GSC_v4.0','sample','n','n','air',',');
INSERT INTO `VARIABLE_NAMES` VALUES (338,'wind speed','speed of wind measured at the time of sampling','{float} {unit}','wind_speed','GSC_v4.0','sample','n','n','air',',');
INSERT INTO `VARIABLE_NAMES` VALUES (339,'alt id - dbgap study id','Name or identifier used in dbGAP to represent the Study/Project to which this sample belongs.','{text}',NULL,'non-gsc','sample','n','n','n/a','added for gap study IDs');
INSERT INTO `VARIABLE_NAMES` VALUES (340,'GOLD sample classification','This is the \"phylogenetic\" equivilent classification proposed and used by GOLD and JGI, http://www.genomesonline.org/cgi-bin/metagenomic_classification_tree.cgi, this should match the one given by GOLD if an \"alt id - GOLD\" is available, if not then it may be classified by EMG but this may or maynot be propogated to GOLD?!','{text}','GOLD_Class','non-gsc','sample','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (341,'NCBI sample classification','This is the \"phylogenetic\" equivilent classification proposed and used by NCBI and SRA (http://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=408169) this should always match the one given in the SRA sample object.','{float}','taxid','GSC_v2.1','sample','y','y','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (342,'alt id - BioSample id','The BioSample database contains descriptions of biological source materials used in experimental assays.','{text}','biosample','non-gsc','sample','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (343,'alt id - BioProject id','A BioProject is a collection of biological data related to a single initiative, originating from a single organization or from a consortium. A BioProject record provides users a single place to find links to the diverse data types generated for that project.','{text}','bioproject','non-gsc','study','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (344,'extreme_unusual_properties/Cu','concentration of Copper (Cu)','{float} {unit}','extreme_unusual_properties/Cu','non-gsc','sample','n','n','all','added for sediment samples ERP001568');
INSERT INTO `VARIABLE_NAMES` VALUES (345,'extreme_unusual_properties/Cd','concentration of Cadmium (Cd)','{float} {unit}','extreme_unusual_properties/Cd','non-gsc','sample','n','n','all','added for sediment samples ERP001568');
INSERT INTO `VARIABLE_NAMES` VALUES (346,'extreme_unusual_properties/Ni','concentration of Nickel (Ni)','{float} {unit}','extreme_unusual_properties/Ni','non-gsc','sample','n','n','all','added for sediment samples ERP001568');
INSERT INTO `VARIABLE_NAMES` VALUES (347,'extreme_unusual_properties/Zn','concentration of Zinc (Zn)','{float} {unit}','extreme_unusual_properties/Zn','non-gsc','sample','n','n','all','added for sediment samples ERP001568');
INSERT INTO `VARIABLE_NAMES` VALUES (348,'extreme_unusual_properties/Pb','concentration of Lead (Pb)','{float} {unit}','extreme_unusual_properties/Pb','non-gsc','sample','n','n','all','added for sediment samples ERP001568');
INSERT INTO `VARIABLE_NAMES` VALUES (349,'extreme_unusual_properties/Mn','concentration of Manganese (Mn)','{float} {unit}','extreme_unusual_properties/Mn','non-gsc','sample','n','n','all','added for sediment samples ERP001568');
INSERT INTO `VARIABLE_NAMES` VALUES (350,'extreme_unusual_properties/Fe','concentration of Iron (Fe)','{float} {unit}','extreme_unusual_properties/Fe','non-gsc','sample','n','n','all','added for sediment samples ERP001568');
INSERT INTO `VARIABLE_NAMES` VALUES (351,'extreme_unusual_properties/As','concentration of Arsenic (As)','{float} {unit}','extreme_unusual_properties/As','non-gsc','sample','n','n','all','added for sediment samples ERP001568');
INSERT INTO `VARIABLE_NAMES` VALUES (352,'instrument model',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (354,'bicarbonate','concentration of bicarbonate','{float} {unit}','bicarbonate','non-gsc','sample','n','n','n/a',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (356,'ENA checklist','ENA checklist accession','{text}','ena_checklist','non-gsc','sample','n','n',NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (357,'sampling cruise',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (358,'sampling site',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (359,'minimum filter size',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (360,'maximum filter size',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (361,'oxygen saturation',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (364,'fluorescence',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (365,'collection time','The time of sampling, either as an instance (single point in time) or interval.',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (366,'photosynthetically active radiation',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (367,'potential density','sigma-theta',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (368,'geographic location (country and/or sea)','The geographical origin of the sample as defined by the country or sea name. Country or sea names should be chosen from the INSDC country list (http://insdc.org/country.html), or the GAZ ontology (v1.446) (http://bioportal.bioontology.org/visualize/40651)','{term}:{term}:{text}','geo_loc_name','GSC_v4.0','sample','y','y','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (369,'geographic location (region and locality)','The geographical origin of the sample: region and locality',NULL,NULL,'GSC_v4.0','sample','n','n','all',NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (370,'event date/time','event date/time',NULL,NULL,'microB3',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (371,'event_device','event_device',NULL,NULL,'microB3',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (372,'latitude end','latitude end',NULL,NULL,'microB3',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (373,'latitude start','latitude start',NULL,NULL,'microB3',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (374,'longitude end','longitude end',NULL,NULL,'microB3',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (375,'longitude start','longitude start',NULL,NULL,'microB3',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (376,'marine region','marine region',NULL,NULL,'microB3',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (377,'protocol label','protocol label',NULL,NULL,'microB3',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (378,'sample_title','sample_title',NULL,NULL,'microB3',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (379,'sampling campaign','sampling campaign',NULL,NULL,'microB3',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (380,'sampling platform','sampling platform',NULL,NULL,'microB3',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (381,'sampling_objective','sampling_objective',NULL,NULL,'microB3',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (382,'absolute air humidity',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (383,'air temperature',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (384,'annotation source',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (385,'bacterial production',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (386,'bacterial respiration',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (387,'building occupancy type',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (388,'building setting',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (389,'conductivity',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (390,'dew point',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (391,'downward PAR',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (392,'encoded traits',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (393,'estimated size',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (394,'extrachromosomal elements',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (395,'filter type',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (396,'finishing strategy',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (397,'health or disease status of specific host',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (398,'heating and cooling system type',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (399,'host HIV status',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (400,'host age',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (401,'host blood pressure diastolic',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (402,'host blood pressure systolic',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (403,'host body habitat',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (404,'host body product',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (405,'host body site',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (406,'host body-mass index',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (407,'host diet',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (408,'host disease status',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (409,'host dry mass',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (410,'host family relationship',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (411,'host genotype',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (412,'host height',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (413,'host infra-specific name',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (414,'host infra-specific rank',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (415,'host last meal',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (416,'host length',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (417,'host life stage',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (418,'host occupation',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (419,'host phenotype',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (420,'host pulse',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (421,'host sex',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (422,'host substrate',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (423,'host total mass',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (424,'host wet mass',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (425,'indoor space',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (426,'indoor surface',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (427,'isolation and growth condition',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (428,'known pathogenicity',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (429,'light type',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (430,'number of replicons',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (431,'observed biotic relationship',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (432,'occupancy at sampling',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (433,'occupant density at sampling',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (434,'ploidy',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (435,'propagation',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (436,'reference for biomaterial',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (437,'relative air humidity',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (438,'sample size sorting method',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (439,'sample volume or weight for DNA extraction',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (440,'space typical state',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (441,'subspecific genetic lineage',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (442,'substructure type',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (443,'surface air contaminant',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (444,'surface humidity',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (445,'surface material',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (446,'surface moisture',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (447,'surface moisture pH',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (448,'surface temperature',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (449,'total nitrogen method',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (450,'trophic level',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (451,'typical occupant density',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (452,'water content method',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (453,'water current',NULL,NULL,NULL,'GSC_v4.0',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (454,'chlorophyll sensor',NULL,NULL,NULL,'TARA Oceans',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (455,'citation',NULL,NULL,NULL,'TARA Oceans',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (456,'event date/time end',NULL,NULL,NULL,'TARA Oceans',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (457,'event date/time start',NULL,NULL,NULL,'TARA Oceans',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (458,'event label',NULL,NULL,NULL,'TARA Oceans',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (459,'further details',NULL,NULL,NULL,'TARA Oceans',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (460,'last update date',NULL,NULL,NULL,'TARA Oceans',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (461,'nitrate sensor',NULL,NULL,NULL,'TARA Oceans',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (462,'oxygen sensor',NULL,NULL,NULL,'TARA Oceans',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (463,'salinity sensor',NULL,NULL,NULL,'TARA Oceans',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (464,'sample collection device',NULL,NULL,NULL,'TARA Oceans',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (465,'sample status',NULL,NULL,NULL,'TARA Oceans',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (466,'sampling station',NULL,NULL,NULL,'TARA Oceans',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (467,'size fraction lower threshold',NULL,NULL,NULL,'TARA Oceans',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (468,'size fraction upper threshold',NULL,NULL,NULL,'TARA Oceans',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (469,'barcode sequence',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (470,'% templated ISPs',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (471,'background RFU (negative control)',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (472,'Bioanalyzer',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (473,'extraction',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (474,'protocol reference',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (475,'run ID',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (476,'sample name',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (477,'source name',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (478,'ammonium nitrogen',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (479,'anonymized name',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (480,'barcode number',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (481,'boron hot CaCl2',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (482,'broker object ID',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (483,'bulk density',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (484,'clay',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (485,'collection site',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (486,'collection timezone',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (487,'colour',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (488,'common name',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (489,'continent',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (490,'coarse sand',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (491,'fine sand',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (492,'developmental stage',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (493,'gravel',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (494,'host',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (495,'isolation source',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (496,'material',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (497,'nitrate nitrogen',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (498,'organism',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (499,'pcr amplification',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (500,'sand',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (501,'silt',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (502,'Colwell phosphorus',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (503,'pH level (CaCl2)',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (504,'pH level (H2O)',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (505,'Colwell potassium',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (506,'sample size',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (507,'sample identifier',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (508,'sulphur',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (509,'tissue site',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (510,'tissue type',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (511,'unenriched ISPs',NULL,NULL,NULL,'non-GSC',NULL,NULL,NULL,NULL,NULL);
INSERT INTO `VARIABLE_NAMES` VALUES (512,'size fraction',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (530,'Sort Day',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (531,'Weight for Height Z-score',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (532,'Length for Age Z score',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (533,'env_material',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (534,'BioSampleModel',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (535,'ref_biomaterial',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (536,'sample_type',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (537,'strain',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (538,'Is the sequenced pathogen host associated?',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (539,'isolate',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (540,'host health state',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (541,'environmental_sample',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (542,'collected_by',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (543,'host scientific name',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (544,'respiration rate',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (545,'o2 consumption',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (546,'geographic location',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (547,'label',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (554,'source_material_id',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (576,'name of the sampling site',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (577,'submitted subject id',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (578,'gap_subject_id',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (579,'analyte type',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (580,'gap_accession',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (581,'gap_consent_short_name',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (582,'submitter handle',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (583,'gap_sample_id',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (584,'gap_consent_code',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (585,'biospecimen repository sample id',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (586,'biospecimen repository',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (587,'study design',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (588,'submitted sample id',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (589,'study name',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (590,'Product Part',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (591,'geographic location (latitude and longitude)',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (592,'Product_number',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (593,'Product_Part',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (594,'Cheese variety',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (595,'Factory',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (596,'Sampling_zone',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (597,'Replicate',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (615,'Sample Number',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (616,'Variation',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (617,'Barcode',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (618,'storage conditions',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (619,'annual and seasonal temperature',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (620,'salinity method',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (621,'pooling of dna extracts',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (622,'fao classification',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (623,'al saturation method',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (624,'local classification',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (625,'annual and seasonal precipitation',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (626,'heavy metals method',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (627,'al saturation',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (628,'local classification method',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (629,'Timepoint',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (630,'Type',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (631,'ethnicity',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (632,'host_tissue_sampled',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (633,'ecotype',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (634,'isol_growth_condt',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (635,'cultivar',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (636,'BMI',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (637,'seq_qual_check',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (638,'Abx',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (639,'Hx_Prev',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (640,'White',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (641,'tissue',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (642,'cancer_stage',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (643,'fit_result',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (644,'Hx_Fam_CRC',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (645,'NSAID',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (646,'Ethnic',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (647,'Black',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (648,'Diabetic',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (649,'Native',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (650,'biomaterial_provider',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (651,'Diabetes_Med',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (652,'seq_methods',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (653,'Asian',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (654,'Hx_of_Polyps',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (655,'Pacific',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (656,'diagnosis',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (657,'Smoke',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (658,'Weight',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (659,'description',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (660,'collection_method',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (661,'DNAtemplate_forPCR',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (662,'hrs_after_initial_sampling',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (663,'sampling_order',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (664,'run_date',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (665,'host_disease',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (666,'biotic_relationship',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (667,'rRNA removal',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (668,'Experimental Conditions',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (669,'Euk mRNA removal',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (670,'sample_title:sample_name',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (671,'unique_smple_id',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (672,'Patient_timepoint',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (673,'sample_code',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (674,'time',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (675,'host_strain',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (676,'week_of_experiment',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (677,'host_ancestral_origin',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (678,'microbiota_donor_strain',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (679,'microbiota_donor_origin',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (680,'host_origin',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (681,'ENA-SUBMISSION-TOOL',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (682,'mouse batch',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (683,'patho',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (684,'detail_pathogen',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (685,'treatment',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (686,'delivery_mode',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (687,'oxalate consumption',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (688,'HIV_viral_load',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (689,'host_sexual_orientation',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (690,'host_Residency_Area',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (691,'host_allergy',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (692,'stool_consistency',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (693,'Leukocytes',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (694,'HIV_Profile',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (695,'PCR_Neisseria_gonorrhoeae',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (696,'host_deposition_frequency_per_day',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (697,'PCR_human_papilloma_virus',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (698,'host_abdominal_transit_alterations',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (699,'HIV_RiskGroup',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (700,'HIV_serostatus',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (701,'Anal_cytology',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (702,'host_other_gender',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (703,'CD4+_Tcell_counts',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (704,'host_country_of_origin',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (705,'Lymphocytes',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (706,'HCV_coinfection',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (707,'Syphilis_serology',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (708,'HBV_coinfection',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (709,'PCR_Chlamydia_trachomatis',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (710,'file_prefix',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (711,'sample_description',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (712,'sample comment',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (793,'title',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (794,'nutrition_sugar',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (795,'age_unit',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (796,'nutrition_fiber',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (797,'\"PUBLIC\"',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (798,'nutrition_fat',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (799,'assigned_from_geo',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (800,'nutrition_carb',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (801,'nutrition_cholesterol',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (802,'nutrition_protein',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (803,'nutrition_satfat',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (804,'illumina_technology',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (805,'collection_day',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (806,'nutrition_calcium',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (807,'nutrition_sodium',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (808,'nutrition_calorie',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (809,'human gut environmental package',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (810,'water environmental package',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (812,'soil environmental package',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (813,'host-associated environmental package',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (814,'subject',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (815,'study',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (816,'air environmental package',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (817,'miscellaneous environmental package',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (819,'human-associated environmental package',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (820,'extraction method',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (821,'health_state',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (822,'sff_name',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (823,'Mouthwash',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (824,'ArrayExpress-Sex',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (825,'ArrayExpress-Species',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (826,'16s variable region',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (827,'sediment environmental package',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (828,'plant-associated environmental package',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (830,'microbial mat/biofilm environmental package',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (831,'Plant age',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (832,'Phosphate level',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (833,'Inoculated',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (834,'Fraction',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (835,'Plant genotype',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (836,'Harvest method',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (837,'Sequencing plate',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (838,'Biological replicate',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (839,'Day',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (840,'wastewater/sludge environmental package',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (841,'Salicylic acid treatment',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (842,'block',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (843,'site',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (844,'sequence_batch',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (845,'plant_id',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (846,'population',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (847,'cohort',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (848,'library',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (849,'LTSP_soil_samples',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (850,'LTSP_treatment',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (851,'Carbon_isotope_label',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (852,'cn',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (853,'total_c',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (854,'moisture_content',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (855,'adaptor sequence',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (856,'climatic_zone',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (857,'harvesting_treatment',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (858,'mean_annual_temperatuere_celsius',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (859,'tree_cover',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (860,'soil_classification',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (861,'second_file_name',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (862,'second_md5_checksum',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (863,'host_species',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (864,'first_md5_checksum',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (865,'first_file_name',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (866,'hpv status',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (867,'hpv status at follow up',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (868,'Geographic location (locality)',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (869,'Colection Date',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (870,'Geographic location (area)',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (871,'Culture origin',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (872,'sample_strategy',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (873,'misc_param_1',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (874,'\"geographic location (country and/or sea,region)\"',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (917,'collection_timestamp',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (918,'zylenes_total',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (919,'cycloalkanes',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (920,'sample_progress',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (921,'toluene',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (922,'required_sample_info_status',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (923,'tph_sum',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (924,'diss_inorg_nitrogen',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (925,'isopropylbenzene',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (926,'diss_phosphate',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (927,'diss_nitrate',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (928,'benzene',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (929,'alkylated_pahs',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (930,'distance',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (931,'dna_extracted',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (932,'physical_specimen_location',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (933,'physical_specimen_remaining',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (934,'tot_sulfur',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (935,'public',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (936,'pah_sum',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
INSERT INTO `VARIABLE_NAMES` VALUES (937,'ethylbenzene',NULL,NULL,NULL,NULL,'sample',NULL,NULL,NULL,'Auto-inserted by webuploader');
/*!40000 ALTER TABLE `VARIABLE_NAMES` ENABLE KEYS */;
UNLOCK TABLES;
-- Dump completed on 2019-09-06 14:23:18
